File: pglite/src/pglite/release/preload/packageLoader.ts
import type {
	DataManifest,
	RuntimeModule,
} from "../core/runtimeTypes.js";
import { applyFsBootstrapPaths } from "../fs/fsBridge.js";
import type { FsPathEntry } from "../core/runtimeTypes.js";
import { DataRequest } from "./dataRequest.js";

function assert(condition: unknown, message: string): asserts condition {
	if (!condition) {
		throw new Error(message);
	}
}

async function fetchRemotePackage(
	module: RuntimeModule,
	packageName: string,
	packageSize: number,
): Promise<ArrayBuffer> {
	module.dataFileDownloads ??= {};
	const response = await fetch(packageName);
	if (!response.ok) {
		throw new Error(`${response.status}: ${response.url}`);
	}
	if (!response.body && response.arrayBuffer) {
		return response.arrayBuffer();
	}
	assert(response.body, `response body is missing: ${packageName}`);
	const reader = response.body.getReader();
	const chunks: Uint8Array[] = [];
	const total = Number(response.headers.get("Content-Length") ?? packageSize);
	let loaded = 0;
	for (;;) {
		const { done, value } = await reader.read();
		if (done) {
			break;
		}
		chunks.push(value);
		loaded += value.length;
		module.dataFileDownloads[packageName] = { loaded, total };
		let totalLoaded = 0;
		let totalSize = 0;
		for (const download of Object.values(module.dataFileDownloads)) {
			totalLoaded += download.loaded;
			totalSize += download.total;
		}
		module.setStatus?.(`Downloading data... (${totalLoaded}/${totalSize})`);
	}
	const packageData = new Uint8Array(chunks.reduce((sum, c) => sum + c.length, 0));
	let offset = 0;
	for (const chunk of chunks) {
		packageData.set(chunk, offset);
		offset += chunk.length;
	}
	return packageData.buffer;
}

function processPackageData(
	module: RuntimeModule,
	manifest: DataManifest,
	arrayBuffer: ArrayBuffer,
): void {
	assert(arrayBuffer, "Loading data file failed.");
	const byteArray = new Uint8Array(arrayBuffer);
	for (const file of manifest.files) {
		const request = new DataRequest(module, byteArray, file);
		request.open(file.filename);
		request.onload();
	}
	module.removeRunDependency?.("datafile_pglite.data");
}

export interface PackageLoaderOptions {
	manifest: DataManifest;
	fsPaths: FsPathEntry[];
	packageName?: string;
	remotePackageBase?: string;
}

export function installPackageLoader(
	module: RuntimeModule,
	options: PackageLoaderOptions,
): void {
	const packageName = options.packageName ?? "pglite.data";
	const remotePackageBase = options.remotePackageBase ?? "pglite.data";
	const runWithFs = async (mod: RuntimeModule): Promise<void> => {
		applyFsBootstrapPaths(mod, options.fsPaths);
		mod.addRunDependency?.("datafile_pglite.data");
		mod.preloadResults ??= {};
		mod.preloadResults[packageName] = { fromCache: false };
		const remotePackageName = mod.locateFile
			? mod.locateFile(remotePackageBase, "")
			: remotePackageBase;
		let fetched = mod.getPreloadedPackage
			? mod.getPreloadedPackage(remotePackageName, options.manifest.remotePackageSize)
			: null;
		if (!fetched) {
			fetched = await fetchRemotePackage(
				mod,
				remotePackageName,
				options.manifest.remotePackageSize,
			);
		}
		processPackageData(mod, options.manifest, fetched);
	};

	const preRun = (mod: RuntimeModule) => {
		void runWithFs(mod).catch((error) => {
			throw error;
		});
	};

	if (module.calledRun) {
		preRun(module);
		return;
	}
	const hooks = (module.preRun ??= []);
	hooks.push(() => preRun(module));
}

File: src/main/java/io/github/hidekatsu_izuno/pglite_jdbc/pglite/release/preload/packageLoader.java
package io.github.hidekatsu_izuno.pglite_jdbc.pglite.release.preload;

import io.github.hidekatsu_izuno.pglite_jdbc.pglite.release.core.runtimeTypes;
import io.github.hidekatsu_izuno.pglite_jdbc.pglite.release.fs.fsBridge;
import io.github.hidekatsu_izuno.pglite_jdbc.pglite.utils;
import java.util.HashMap;
import java.util.Map;

public class packageLoader {
    public record PackageLoaderOptions(
        runtimeTypes.DataManifest manifest,
        java.util.List<runtimeTypes.FsPathEntry> fsPaths,
        String packageName,
        String remotePackageBase
    ) {}

    private packageLoader() {}

    public static void installPackageLoader(runtimeTypes.RuntimeModule module, PackageLoaderOptions options) {
        var packageName = options.packageName() != null ? options.packageName() : "pglite.data";
        var base = options.remotePackageBase() != null ? options.remotePackageBase() : packageName;
        Runnable runWithFs = () -> {
            fsBridge.applyFsBootstrapPaths(module, options.fsPaths());
            module.addRunDependency("datafile_pglite.data");

            var preloadResults = module.preloadResults();
            if (preloadResults == null) {
                preloadResults = new HashMap<>();
                module.setPreloadResults(preloadResults);
            }
            preloadResults.put(packageName, Map.of("fromCache", Boolean.FALSE));

            var remotePackageName = module.locateFile(base, "");
            var fetched = module.getPreloadedPackage(remotePackageName, options.manifest().remotePackageSize());
            if (fetched == null) {
                fetched = fetchRemotePackage(module, remotePackageName, options.manifest().remotePackageSize());
            }
            processPackageData(module, options.manifest(), fetched);
        };

        if (module.calledRun()) {
            runWithFs.run();
            return;
        }
        var hooks = new java.util.ArrayList<>(module.preRun());
        hooks.add(runWithFs);
        module.setPreRun(hooks);
    }

    private static void processPackageData(
        runtimeTypes.RuntimeModule module,
        runtimeTypes.DataManifest manifest,
        byte[] bytes
    ) {
        if (bytes == null) {
            throw new IllegalStateException("Loading data file failed.");
        }
        for (var file : manifest.files()) {
            var request = new dataRequest(module, bytes, file);
            request.open(file.filename());
            request.onload();
        }
        module.removeRunDependency("datafile_pglite.data");
    }

    private static byte[] fetchRemotePackage(
        runtimeTypes.RuntimeModule module,
        String packageName,
        int packageSize
    ) {
        var downloads = module.dataFileDownloads();
        if (downloads == null) {
            downloads = new HashMap<>();
            module.setDataFileDownloads(downloads);
        }
        var bytes = utils.readFile(packageName);
        var total = bytes.length > 0 ? bytes.length : packageSize;
        downloads.put(packageName, Map.of("loaded", bytes.length, "total", total));
        module.setStatus("Downloading data... (" + bytes.length + "/" + total + ")");
        return bytes;
    }
}
